### PR TITLE
Automate population of filament & hms error consts via scripts

### PR DESCRIPTION
**Pull request by t3chguy**
Would prevent the need for PRs like https://github.com/greghesp/ha-bambulab/pull/554 & https://github.com/greghesp/ha-bambulab/pull/553 in future

Fixes https://github.com/greghesp/ha-bambulab/issues/567

### CODE REVIEW

File: scripts/fetch_filaments.py [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+def format_filament_name(name: str) -> str:
+    if " @" in name:
+        return name[:name.index(" @")]
+    return name
+
+
+filaments = get_filaments()
+data = dict(map(lambda filament: (filament["filament_id"], format_filament_name(filament["name"])), filaments))
+
+print(json.dumps(data, indent=4))
```
Comment (maruel) [inline scripts/fetch_filaments.py:28]: Use `sort_keys=True` so the file is consistent across runs.

File: scripts/fetch_filaments.py [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+
+
+def format_filament_name(name: str) -> str:
+    if " @" in name:
+        return name[:name.index(" @")]
+    return name
+
+
+filaments = get_filaments()
+data = dict(map(lambda filament: (filament["filament_id"], format_filament_name(filament["name"])), filaments))
```
Comment (maruel) [inline scripts/fetch_filaments.py:26]: Here's a much shorter version:
```
data = {f["filament_id"]: f["name"].split("@", 1)[0] for f in get_filaments()}
```

File: scripts/fetch_filaments.py [hunk 3]
```diff
@@ -0,0 +1,28 @@
+from __future__ import annotations
+
+import json
+import requests
+
+"""
+python3 scripts/fetch_filaments.py > custom_components/bambu_lab/pybambu/filaments.json
```
Comment (maruel) [inline scripts/fetch_filaments.py:7]: Add before the imports.

File: scripts/fetch_filaments.py [hunk 4]
```diff
@@ -0,0 +1,28 @@
+from __future__ import annotations
```
Comment (maruel) [inline scripts/fetch_filaments.py:1]: Unnecessary since HA doesn't support old python versions.

File: scripts/fetch_filaments.py [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+
+def get_filaments() -> dict:
+    url = 'https://api.bambulab.com/v1/iot-service/api/slicer/setting?version=undefined'
+    response = requests.get(url, timeout=10)
+    if not response.ok:
+        raise ValueError(response.status_code)
+    return response.json()["filament"]["public"]
+
+
+def format_filament_name(name: str) -> str:
```
Comment (maruel) [inline scripts/fetch_filaments.py:19]: Unnecessary

**Review by maruel — APPROVED** [review]
Thanks! A few stylistic improvements.

File: scripts/fetch_filaments.py (hunk 4 above)
Comment (t3chguy) [reply scripts/fetch_filaments.py:1]: image

I copied it from the other files which use it

File: scripts/fetch_filaments.py (hunk 2 above)
Comment (t3chguy) [reply scripts/fetch_filaments.py:26]: This leaves a trailing space, will add a trim

**Review by andybateman — APPROVED** [review]

**Review by promiseofcake — APPROVED** [review]

**Comment by Kirbo** [comment]
[This issue](https://github.com/greghesp/ha-bambulab/issues/567) is still waiting for this PR to be merged.

**Pull request merged**